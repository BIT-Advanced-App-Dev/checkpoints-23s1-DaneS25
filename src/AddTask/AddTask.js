/* 
Add Task component.
Click add task to add a new title and description of task
*/
import Modal from "../Modal/Modal";
import { useState } from "react";
import "./addTask.css";
import { db } from "../firebase"; // Importing the Firebase database configuration
import {
  serverTimestamp,
  doc,
  collection,
  addDoc,
  getDoc
} from "firebase/firestore"; // Importing Firebase Firestore functionalities

function AddTask({ onClose, open, userUid }) {
  // Create state variables for title and description of the tasks
  const [title, setTitle] = useState('')
  const [description, setDescription] = useState('')
  
  const handleSubmit = async (e) => {
    e.preventDefault();
  
    try {
      const userDocRef = doc(db, "users", userUid);
      const tasksCollectionRef = collection(userDocRef, "tasks");
      const newTaskDocRef = await addDoc(tasksCollectionRef, {
        uid: userUid,
        title: title,
        description: description,
        completed: false,
        created: serverTimestamp(),
      });
      console.log("New task created with ID: ", newTaskDocRef.id);
  
      // Get the new task document with taskId
      const taskDocRef = doc(db, "users", userUid, "tasks", newTaskDocRef.id);
      const taskDoc = await getDoc(taskDocRef);
      console.log("Task data: ", taskDoc.data());
  
      onClose();
    } catch (error) {
      console.error("Error adding document: ", error);
    }
  };

  return (
    <Modal modalLabel="Add Task" onClose={onClose} open={open}>
      {/* Render in the Modal component with props modalLabel, onClose, and open */}
      <form onSubmit={handleSubmit} className="addTask" name="addTask">
        <input
          type="text"
          name="title"
          onChange={(e) => setTitle(e.target.value.toUpperCase())}
          value={title}
          placeholder="Enter title"
        />
        <textarea
          name="description"
          onChange={(e) => setDescription(e.target.value)}
          placeholder="Enter task description"
          value={description}
        ></textarea>
        <button type="submit" name="addTask">
          Done
        </button>
        {/* Button that will trigger the handleSubmit function when clicked */}
      </form>
    </Modal>
  );
}

export default AddTask


