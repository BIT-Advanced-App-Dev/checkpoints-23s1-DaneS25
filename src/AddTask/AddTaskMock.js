import Modal from "../Modal/Modal"
import { useState } from 'react'
import './addTask.css'

function AddTaskMock({ onClose, open, mockAddDoc }) {
  const [title, setTitle] = useState('')
  const [description, setDescription] = useState('')

  const handleSubmit = async (e) => {
    e.preventDefault();
  
    try {
      await mockAddDoc({
        title: title,
        description: description,
        completed: false,
        created: { seconds: 1234567890 }
      });
  
      onClose();
    } catch (err) {
        console.log(err);
      console.error(err);
    }
  };
  

  return (
    <Modal modalLabel='Add Task' onClose={onClose} open={open}>
      <form onSubmit={handleSubmit} className='addTask' name='addTask'>
        <input
          type='text'
          name='title'
          onChange={(e) => setTitle(e.target.value.toUpperCase())}
          value={title}
          placeholder='Enter title'
        />
        <textarea
          name='description'
          onChange={(e) => setDescription(e.target.value)}
          placeholder='Enter task description'
          value={description}
        ></textarea>
        <button type='submit' name='done'>
          Done
        </button>
      </form>
    </Modal>
  )
}

export default AddTaskMock



  