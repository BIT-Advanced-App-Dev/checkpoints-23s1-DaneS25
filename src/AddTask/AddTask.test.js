import React from 'react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import AddTask from './AddTask';

it('renders AddTask component correctly', () => {
  const onClose = jest.fn(); // Create a mock function for the onClose prop
  const component = renderer.create(<AddTask onClose={onClose} open={true} />); // Create the component
  let tree = component.toJSON(); // Get the initial snapshot

  // Simulate user input
  act(() => {
    const titleInput = component.root.findByProps({ name: 'title' }); // Find the title input element
    titleInput.props.onChange({ target: { value: 'Test Title' } }); // Update the title state variable
    const descriptionInput = component.root.findByProps({ name: 'description' }); // Find the description input element
    descriptionInput.props.onChange({ target: { value: 'Test Description' } }); // Update the description state variable
  });

  // Get a new snapshot after the state variables have been updated
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // Simulate form submission
  act(() => {
    const form = component.root.findByProps({ name: 'addTask' }); // Find the form element
    form.props.onSubmit({ preventDefault: jest.fn() }); // Submit the form with a mock event object
  });
});


