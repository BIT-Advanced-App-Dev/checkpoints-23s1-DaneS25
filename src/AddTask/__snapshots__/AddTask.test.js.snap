// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AddTask component correctly 1`] = `
<div
  className="modalContainer"
  onClick={[Function]}
>
  <div
    className="modal undefined"
  >
    <div
      className="modal__head"
    >
      <h2>
        Add Task
      </h2>
      <span
        className="modal__close"
        onClick={[MockFunction]}
      >
        x
      </span>
    </div>
     
    <form
      className="addTask"
      name="addTask"
      onSubmit={[Function]}
    >
      <input
        name="title"
        onChange={[Function]}
        placeholder="Enter title"
        type="text"
        value="TEST TITLE"
      />
      <textarea
        name="description"
        onChange={[Function]}
        placeholder="Enter task description"
        value="Test Description"
      />
      <button
        name="addTask"
        type="submit"
      >
        Done
      </button>
       
    </form>
  </div>
</div>
`;
